### ISSUE TITLE
Support making HTTP requests through unix sockets

### ISSUE THREAD

**Issue by smcassar** [dart-lang/sdk#42716]
Dart Version: Dart VM version: 2.8.4 (stable) on "macos_x64"

I'd like to interact with the Docker Machine API which exposes a restful api through a unix socket but this currently doesn't look possible.

I can currently able connect to the socket using `Socket.connect` but there doesn't appear to be a way to use any of the HTTP classes to format the request or process the response.

I've also asked this question on Gitter but we weren't able to find a solution.

If this is already possible then an example would be highly appreciated.

**Comment by sortie** [comment]
I'm thinking the cleanest solution is having HttpClient support the `http+unix://` URI scheme (is there any standard reference on that name?). The `HttpClient` class would support it much it would support `http` and `https`.

There may be some security considerations if anything is unquestionably forwarding requests to a HttpClient under the assumption that the scheme is always http or https.

Alternatively, or additionally, the HttpClient class could have a way of letting you inject your own connections. I'm not entirely sure if that's already possible via the proxy or HttpOverrides API. Potentially some overrides API could let you establish the connections yourself.

@zichangg @lrhn What do you think?

**Comment by zichangg** [comment]
> I'm thinking the cleanest solution is having HttpClient support the http+unix:// URI scheme (is there any standard reference on that name?). The HttpClient class would support it much it would support http and https.

Lgtm. The only question is whether https should be applied together with unix domain socket. Servers and clients are on the same host machine. Is ssl still required? Maybe some testing will use this feature. If it is needed, class `secureSocket` should also be modified accordingly.

> There may be some security considerations if anything is unquestionably forwarding requests to a HttpClient under the assumption that the scheme is always http or https.

I don't think http_unix:// URI scheme violates the assumption. Do you have an example? The higher level protocol remains unchanged. It is the underlying connection that has been substituted.

**Comment by sortie** [comment]
https doesn't make sense for Unix sockets. They are always local and impossible to intercept and also impossible to authenticate.

That is `http+unix`, not `http_unix` btw. My concern is rather programs that erroneously allow something from another trust domain to request any URI (e.g. think a proxy, or a Flutter app that's a social network and allows people to link to stuff, and it prefetches stuff, and now people's viewing posts causes an access to an unix socket without consent, or stuff like that), and assumes HttpClient only supports `http` and `https` and such, and not unix socket communication and other schemes. There's been security issues like that in the past where something supported more protocols than useful and it ended up being surprising when they were abused, that's e.g. something browsers needs to be super careful about. I'm just thinking aloud so we always have security in our mind.

**Comment by zichangg** [comment]
@smcassar I took a look at our Overrides API. You can use `IOOverride` as a workaround.
https://api.dart.dev/stable/2.8.4/dart-io/IOOverrides/socketStartConnect.html

Our `HttpClient` uses `Socket.startConnect` to build the connection. https://github.com/dart-lang/sdk/blob/bbe6bb9823495574917c75634b1f8a2a8099086c/sdk/lib/_http/http_impl.dart#L2076

Here is [an example](https://api.dart.dev/stable/2.8.4/dart-io/IOOverrides-class.html) of how to use `IOOverride`.

then you can override `socketStartConnect` to create a unix domain socket,
```dart
Future<ConnectionTask<Socket>> socketStartConnect(host, int port,
    {sourceAddress}) {
  // Establish a uds socket connection.
  // Make sure `host` is an `InternetAddress` representing a unix domain address.
  return Socket.startConnect(host, port);
}
```

The final step is to call `HttpClient` APIs within `IOOverrides.runZoned`.

**Comment by zichangg** [comment]
@sortie A great point!! `http+unix` may not be as useful as `http` or `https`. But we do need a better way to support this feature. Personally, I don't like the way utilizing class `IOOverride` which is not obvious or easy for users to use.

We have already had our class `Socket` supporting unix domain sockets. Our `HttpClient` can be consistent with `Socket`. In terms of the cost, this change is not going to be a breaking change. Wdyt?

**Comment by sortie** [comment]
@zichangg Awesome that you found a solution with IOOverride.

Yes, I think we should go ahead with http+unix:// support in HttpClient as it is slightly useful and consistent and is a non-breaking change.

**Comment by robert-ancell** [comment]
As pointed out `http+unix` might be unsafe if you are launching arbitrary URIs. Other methods to consider are:
- Adding an option to HttpClient to always connect to a given unix domain socket instead of using the hostname. This is [what curl uses](https://curl.haxx.se/libcurl/c/CURLOPT_UNIX_SOCKET_PATH.html). The convention is to connect using `http://localhost/some/thing` but the host could be anything, e.g. `http://foo/some/thing`.
- Allowing the user to override the transport in HttpClient. This is [what Go uses](https://golang.org/pkg/net/http/#Client). It's harder than the above method but allows you to support more exotic cases.

**Comment by robert-ancell** [comment]
I've made an [HttpUnixClient](https://github.com/canonical/snapd.dart/blob/master/lib/src/http_unix_client.dart) class that uses the [http package](https://pub.dev/packages/http). It might be of some use to anyone until `dart:io` supports this directly.

**Comment by lrhn** [comment]
I think I'm with @robert-ancell on adding an option to `HttpClient` (or adding a separate `SocketHttpClient`, so we won't have to  change the API of `HttpClient`).

Introducing an abstraction over the transport layer is probably something which will some amount of design, but it is an option.
It would probably be a constructor parameter for `HttpClient`.

**Comment by zichangg** [comment]
Had some discussions offline.

The better way is to have `HttpClient` accept an established `Socket` (It can be unix domain socket!). It will be more flexible to use and less likely to suffer from restrictions of class `URI`.

I opened an issue: https://github.com/dart-lang/sdk/issues/43277

**Comment by cbracken** [comment]
@a-siva and I discussed this last week and he plans to find someone on the team to take a look.

**Comment by a-siva** [comment]
@brianquinlan is going to start looking at this, he will reach out to you @cbracken

**Comment by a-siva** [comment]
I believe https://github.com/dart-lang/sdk/issues/43277 is also a related issue.

**Comment by brianquinlan** [comment]
I have a design proposal to tackle this issue. Any feedback would be welcome:
https://docs.google.com/document/d/1W8TsGXQheNrxQSulr1S3BscZ_ncUsXuPhpAlTsp6lsU/edit#heading=h.xgjl2srtytjt

**Comment by robert-ancell** [comment]
> I have a design proposal to tackle this issue. Any feedback would be welcome: https://docs.google.com/document/d/1W8TsGXQheNrxQSulr1S3BscZ_ncUsXuPhpAlTsp6lsU/edit#heading=h.xgjl2srtytjt

Thanks for working on this @brianquinlan, the proposal looks great. You seem to have come to the same conclusions I did regarding the alternative approaches, and the proposed solution seems the best.

**Comment by natebosch** [comment]
> I don't understand why there are two (official?) HTTP implementations, I guess it's because some core features require HTTP but can't depend on an external package?

`package:http` is not an independent implementation of the protocol, it is a facade on top of the HTTP implementation provided by the SDK which allows using the same code across platforms.

We do have a separate HTTP implementation in a package: https://pub.dev/packages/http2

**Comment by cbracken** [comment]
@a-siva @brianquinlan any updates on progress here?

**Comment by brianquinlan** [comment]
Hey @cbracken ,

We are waiting for a resolution to the breaking change request: https://github.com/dart-lang/sdk/issues/47887

**Comment by brianquinlan** [comment]
Fixed in a0aeed9faad4c793c7679db3b8eed58ab1be6d5c
